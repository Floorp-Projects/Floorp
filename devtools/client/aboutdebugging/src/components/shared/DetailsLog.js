/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

"use strict";

const {
  createFactory,
  PureComponent,
} = require("devtools/client/shared/vendor/react");
const dom = require("devtools/client/shared/vendor/react-dom-factories");
const PropTypes = require("devtools/client/shared/vendor/react-prop-types");

const FluentReact = require("devtools/client/shared/vendor/fluent-react");
const Localized = createFactory(FluentReact.Localized);

const {
  MESSAGE_LEVEL,
} = require("devtools/client/aboutdebugging/src/constants");

/**
 * This component is designed to wrap a warning / error log message
 * in the details tag to hide long texts and make the message expendable
 * out of the box.
 */
class DetailsLog extends PureComponent {
  static get propTypes() {
    return {
      children: PropTypes.node.isRequired,
      type: PropTypes.string,
    };
  }
  getLocalizationString() {
    const { type } = this.props;

    switch (type) {
      case MESSAGE_LEVEL.WARNING:
        return "about-debugging-message-details-label-warning";
      case MESSAGE_LEVEL.ERROR:
        return "about-debugging-message-details-label-error";
      default:
        return "about-debugging-message-details-label";
    }
  }

  render() {
    const { children } = this.props;

    return dom.details(
      {
        className: "details--log",
      },
      Localized(
        {
          id: this.getLocalizationString(),
        },
        dom.summary({}, this.getLocalizationString())
      ),
      children
    );
  }
}

module.exports = DetailsLog;
