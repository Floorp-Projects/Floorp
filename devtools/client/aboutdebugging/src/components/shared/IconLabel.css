/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

.icon-label {
  display: flex;
  column-gap: var(--base-unit);
  align-items: center;
  margin: calc(var(--base-unit) * 2) 0;
  -moz-context-properties: fill;

  font-size: var(--icon-label-font-size);
}

.icon-label--ok {
  --icon-color: var(--icon-ok-color);
}
.icon-label--info {
  --icon-color: var(--icon-info-color);
}

.icon-label__icon {
  padding: var(--base-unit);
  fill: var(--icon-color);
  width: calc(var(--base-unit) * 4);
  height: calc(var(--base-unit) * 4);
}
