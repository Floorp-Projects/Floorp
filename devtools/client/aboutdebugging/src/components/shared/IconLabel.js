/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

"use strict";

const { PureComponent } = require("devtools/client/shared/vendor/react");
const dom = require("devtools/client/shared/vendor/react-dom-factories");
const PropTypes = require("devtools/client/shared/vendor/react-prop-types");

const {
  ICON_LABEL_LEVEL,
} = require("devtools/client/aboutdebugging/src/constants");

const ICONS = {
  [ICON_LABEL_LEVEL.INFO]: "chrome://devtools/skin/images/info.svg",
  [ICON_LABEL_LEVEL.OK]: "chrome://devtools/skin/images/check.svg",
};

/* This component displays an icon accompanied by some content. It's similar
   to a message, but it's not interactive */
class IconLabel extends PureComponent {
  static get propTypes() {
    return {
      children: PropTypes.node.isRequired,
      className: PropTypes.string,
      level: PropTypes.oneOf(Object.values(ICON_LABEL_LEVEL)).isRequired,
    };
  }

  render() {
    const { children, className, level } = this.props;
    return dom.span(
      {
        className: `icon-label icon-label--${level} ${className || ""}`,
      },
      dom.img({
        className: "icon-label__icon",
        src: ICONS[level],
      }),
      children
    );
  }
}

module.exports = IconLabel;
