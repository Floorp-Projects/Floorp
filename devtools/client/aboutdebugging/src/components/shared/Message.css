/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

.message--level-error {
  --message-background-color: var(--error-background);
  --message-border-color: var(--error-border);
  --message-color: var(--error-text);
  --message-icon-color: var(--error-icon);
}

.message--level-info {
  --message-background-color: var(--grey-20);
  --message-border-color: transparent;
  --message-color: var(--grey-90);
  --message-icon-color: var(--grey-90);
}

.message--level-warning {
  --message-background-color: var(--warning-background);
  --message-border-color: var(--warning-border);
  --message-color: var(--warning-text);
  --message-icon-color: var(--warning-icon);
}

/*
 * Layout of the message
 *
 *  +--------+-----------------+----------+
 *  | Icon   | Message content | closing  |
 *  |        | (several lines) | button   |
 *  |        | (     ...     ) |(optional)|
 *  +--------+-----------------+----------+
 */
.message {
  background-color: var(--message-background-color);
  border: 1px solid var(--message-border-color);
  border-radius: var(--base-unit);
  color: var(--message-color);
  display: grid;
  grid-column-gap: var(--base-unit);
  grid-template-columns: calc(var(--base-unit) * 6) 1fr auto;
  grid-template-areas:
    "icon body button";
  margin: calc(var(--base-unit) * 2) 0;
  padding: var(--base-unit);
  -moz-context-properties: fill;
}

.message__icon {
  margin: var(--base-unit);
  fill: var(--message-icon-color);
  grid-area: icon;
}

.message__body {
  align-self: center;
  font-size: var(--message-font-size);
  font-weight: 400;
  grid-area: body;
  line-height: 1.6;
}

.message__button {
  grid-area: button;
  fill: var(--message-icon-color);
}

.message__button:hover {
  /* reverse colors with icon when hover state */
  background-color: var(--message-icon-color);
  fill: var(--message-background-color);
}

.message__button:active {
  /* reverse colors with text when active state */
  background-color: var(--message-color);
  fill: var(--message-background-color);
}
