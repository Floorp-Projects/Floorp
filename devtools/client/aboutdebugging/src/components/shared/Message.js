/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

"use strict";

const {
  createFactory,
  PureComponent,
} = require("devtools/client/shared/vendor/react");
const dom = require("devtools/client/shared/vendor/react-dom-factories");
const PropTypes = require("devtools/client/shared/vendor/react-prop-types");

const FluentReact = require("devtools/client/shared/vendor/fluent-react");
const Localized = createFactory(FluentReact.Localized);

const {
  MESSAGE_LEVEL,
} = require("devtools/client/aboutdebugging/src/constants");

const ICONS = {
  [MESSAGE_LEVEL.ERROR]:
    "chrome://devtools/skin/images/aboutdebugging-error.svg",
  [MESSAGE_LEVEL.INFO]:
    "chrome://devtools/skin/images/aboutdebugging-information.svg",
  [MESSAGE_LEVEL.WARNING]: "chrome://devtools/skin/images/alert.svg",
};
const CLOSE_ICON_SRC = "chrome://devtools/skin/images/close.svg";

/**
 * This component is designed to display a photon-style message bar. The component is
 * responsible for displaying the message container with the appropriate icon. The content
 * of the message should be passed as the children of this component.
 */
class Message extends PureComponent {
  static get propTypes() {
    return {
      children: PropTypes.node.isRequired,
      className: PropTypes.string,
      isCloseable: PropTypes.bool,
      level: PropTypes.oneOf(Object.values(MESSAGE_LEVEL)).isRequired,
    };
  }

  constructor(props) {
    super(props);
    this.state = {
      isClosed: false,
    };
  }

  closeMessage() {
    this.setState({ isClosed: true });
  }

  renderButton(level) {
    return dom.button(
      {
        className:
          `ghost-button message__button message__button--${level} ` +
          `qa-message-button-close-button`,
        onClick: () => this.closeMessage(),
      },
      Localized(
        {
          id: "about-debugging-message-close-icon",
          attrs: {
            alt: true,
          },
        },
        dom.img({
          className: "qa-message-button-close-icon",
          src: CLOSE_ICON_SRC,
        })
      )
    );
  }

  render() {
    const { children, className, level, isCloseable } = this.props;
    const { isClosed } = this.state;

    if (isClosed) {
      return null;
    }

    return dom.aside(
      {
        className:
          `message message--level-${level}  qa-message` +
          (className ? ` ${className}` : ""),
      },
      dom.img({
        className: "message__icon",
        src: ICONS[level],
      }),
      dom.div(
        {
          className: "message__body",
        },
        children
      ),
      // if the message is closeable, render a closing button
      isCloseable ? this.renderButton(level) : null
    );
  }
}

module.exports = Message;
