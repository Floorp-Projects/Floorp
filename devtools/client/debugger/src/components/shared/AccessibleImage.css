/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

.img {
  display: inline-block;
  width: 16px;
  height: 16px;
  vertical-align: middle;
  /* use background-color for the icon color, and mask-image for its shape */
  background-color: var(--theme-icon-color);
  mask-size: contain;
  mask-repeat: no-repeat;
  mask-position: center;
  /* multicolor icons use background-image */
  background-position: center;
  background-repeat: no-repeat;
  background-size: contain;
  /* do not let images shrink when used as flex children */
  flex-shrink: 0;
}

/* Expand arrow icon */
.img.arrow {
  width: 10px;
  height: 10px;
  mask-image: url(chrome://devtools/content/debugger/images/arrow.svg);
  /* we may override the width/height in specific contexts to make the
     clickable area bigger, but we should always keep the mask size 10x10 */
  mask-size: 10px 10px;
  background-color: var(--theme-icon-dimmed-color);
  transform: rotate(-90deg);
  transition: transform 180ms var(--animation-curve);
}

.img.arrow:dir(rtl) {
  transform: rotate(90deg);
}

.img.arrow.expanded {
  /* icon should always point to the bottom (default) when expanded,
  regardless of the text direction */
  transform: none !important;
}

.img.arrow-down {
  mask-image: url(chrome://devtools/content/debugger/images/arrow-down.svg);
}

.img.arrow-up {
  mask-image: url(chrome://devtools/content/debugger/images/arrow-up.svg);
}

.img.blackBox {
  mask-image: url(chrome://devtools/content/debugger/images/blackBox.svg);
}

.img.breadcrumb {
  mask-image: url(chrome://devtools/content/debugger/images/breadcrumbs-divider.svg);
}

.img.case-match {
  mask-image: url(chrome://devtools/content/debugger/images/case-match.svg);
}

.img.close {
  mask-image: url(chrome://devtools/skin/images/close.svg);
}

.img.disable-pausing {
  mask-image: url(chrome://devtools/content/debugger/images/disable-pausing.svg);
}

.img.enable-pausing {
  mask-image: url(chrome://devtools/content/debugger/images/enable-pausing.svg);
  background-color: var(--theme-icon-checked-color);
}

.img.globe {
  mask-image: url(chrome://devtools/content/debugger/images/globe.svg);
}

.img.globe-small {
  mask-image: url(chrome://devtools/content/debugger/images/globe-small.svg);
  mask-size: 12px 12px;
}

.img.window {
  mask-image: url(chrome://devtools/content/debugger/images/window.svg);
}

.img.file {
  mask-image: url(chrome://devtools/content/debugger/images/file-small.svg);
  mask-size: 12px 12px;
}

.img.folder {
  mask-image: url(chrome://devtools/content/debugger/images/folder.svg);
}

.img.home {
  mask-image: url(chrome://devtools/content/debugger/images/home.svg);
}

.img.info {
  mask-image: url(chrome://devtools/skin/images/info.svg);
}

.img.loader {
  mask-image: url(chrome://devtools/content/debugger/images/loader.svg);
}

.img.more-tabs {
  mask-image: url(chrome://devtools/content/debugger/images/command-chevron.svg);
}

html[dir="rtl"] .img.more-tabs {
  transform: scaleX(-1);
}

.img.next {
  mask-image: url(chrome://devtools/content/debugger/images/next.svg);
}

.img.next-circle {
  mask-image: url(chrome://devtools/content/debugger/images/next-circle.svg);
}

.img.pane-collapse {
  mask-image: url(chrome://devtools/content/debugger/images/pane-collapse.svg);
}

.img.pane-expand {
  mask-image: url(chrome://devtools/content/debugger/images/pane-expand.svg);
}

.img.pause {
  mask-image: url(chrome://devtools/content/debugger/images/pause.svg);
}

.img.plus {
  mask-image: url(chrome://devtools/content/debugger/images/plus.svg);
}

.img.prettyPrint {
  mask-image: url(chrome://devtools/content/debugger/images/prettyPrint.svg);
}

.img.refresh {
  mask-image: url(chrome://devtools/content/debugger/images/reload.svg);
}

.img.regex-match {
  mask-image: url(chrome://devtools/content/debugger/images/regex-match.svg);
}

.img.resume {
  mask-image: url(chrome://devtools/content/debugger/images/resume.svg);
}

.img.search {
  mask-image: url(chrome://devtools/content/debugger/images/search.svg);
}

.img.shortcuts {
  mask-image: url(chrome://devtools/content/debugger/images/help.svg);
}

.img.spin {
  animation: spin 0.5s linear infinite;
}

.img.stepIn {
  mask-image: url(chrome://devtools/content/debugger/images/stepIn.svg);
}

.img.stepOut {
  mask-image: url(chrome://devtools/content/debugger/images/stepOut.svg);
}

.img.stepOver {
  mask-image: url(chrome://devtools/content/debugger/images/stepOver.svg);
}

.img.tab {
  mask-image: url(chrome://devtools/content/debugger/images/tab.svg);
}

.img.whole-word-match {
  mask-image: url(chrome://devtools/content/debugger/images/whole-word-match.svg);
}

.img.worker {
  mask-image: url(chrome://devtools/content/debugger/images/worker.svg);
}
