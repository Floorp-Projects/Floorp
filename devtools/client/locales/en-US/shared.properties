# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.

# LOCALIZATION NOTE (dimensions): This is used to display the dimensions
# of a node or image, like 100Ã—200.
dimensions=%S\u00D7%S

# LOCALIZATION NOTE (boxModelSize.accessibleLabel): This is used to read the
# dimensions of a node by a screen reader. This helps communicate
# the visual information in a more explicit form. Example:
# Size: Width 100, height 200.
boxModelSize.accessibleLabel=Size: Width %1$S, height %2$S

# LOCALIZATION NOTE (boxModelInfo.accessibleLabel): This is used to read the
# dimensions and position of a node by a screen reader. This helps communicate
# the visual information in a more explicit form. Example:
# Dimensions: Width 100, height 200, position static.
boxModelInfo.accessibleLabel=Dimensions: Width %1$S, height %2$S, position %3$S

# LOCALIZATION NOTE (boxModelEditable.accessibleLabel): The string spoken by
# screen readers for each button in the box model view that opens that property
# for editing. %1$S is the property displayed in the tooltip when hovering.
# %2$S is the value that is visually displayed.
# Example: margin-left: 0.
boxModelEditable.accessibleLabel=%1$S: %2$S
