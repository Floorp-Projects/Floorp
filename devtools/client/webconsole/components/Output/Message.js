/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

"use strict";

// React & Redux
const {
  Component,
  createFactory,
  createElement,
} = require("devtools/client/shared/vendor/react");
const dom = require("devtools/client/shared/vendor/react-dom-factories");
const { l10n } = require("devtools/client/webconsole/utils/messages");
const actions = require("devtools/client/webconsole/actions/index");
const {
  MESSAGE_LEVEL,
  MESSAGE_SOURCE,
  MESSAGE_TYPE,
} = require("devtools/client/webconsole/constants");
const {
  MessageIndent,
} = require("devtools/client/webconsole/components/Output/MessageIndent");
const MessageIcon = require("devtools/client/webconsole/components/Output/MessageIcon");
const FrameView = createFactory(
  require("devtools/client/shared/components/Frame")
);

loader.lazyRequireGetter(
  this,
  "CollapseButton",
  "devtools/client/webconsole/components/Output/CollapseButton"
);
loader.lazyRequireGetter(
  this,
  "MessageRepeat",
  "devtools/client/webconsole/components/Output/MessageRepeat"
);
loader.lazyRequireGetter(
  this,
  "PropTypes",
  "devtools/client/shared/vendor/react-prop-types"
);
loader.lazyRequireGetter(
  this,
  "SmartTrace",
  "devtools/client/shared/components/SmartTrace"
);

class Message extends Component {
  static get propTypes() {
    return {
      open: PropTypes.bool,
      collapsible: PropTypes.bool,
      collapseTitle: PropTypes.string,
      onToggle: PropTypes.func,
      source: PropTypes.string.isRequired,
      type: PropTypes.string.isRequired,
      level: PropTypes.string.isRequired,
      indent: PropTypes.number.isRequired,
      inWarningGroup: PropTypes.bool,
      isBlockedNetworkMessage: PropTypes.bool,
      topLevelClasses: PropTypes.array.isRequired,
      messageBody: PropTypes.any.isRequired,
      repeat: PropTypes.any,
      frame: PropTypes.any,
      attachment: PropTypes.any,
      stacktrace: PropTypes.any,
      messageId: PropTypes.string,
      scrollToMessage: PropTypes.bool,
      exceptionDocURL: PropTypes.string,
      request: PropTypes.object,
      dispatch: PropTypes.func,
      timeStamp: PropTypes.number,
      timestampsVisible: PropTypes.bool.isRequired,
      serviceContainer: PropTypes.shape({
        emitForTests: PropTypes.func.isRequired,
        onViewSource: PropTypes.func.isRequired,
        onViewSourceInDebugger: PropTypes.func,
        onViewSourceInStyleEditor: PropTypes.func,
        openContextMenu: PropTypes.func.isRequired,
        openLink: PropTypes.func.isRequired,
        sourceMapURLService: PropTypes.any,
      }),
      notes: PropTypes.arrayOf(
        PropTypes.shape({
          messageBody: PropTypes.string.isRequired,
          frame: PropTypes.any,
        })
      ),
      maybeScrollToBottom: PropTypes.func,
      message: PropTypes.object.isRequired,
    };
  }

  static get defaultProps() {
    return {
      indent: 0,
    };
  }

  constructor(props) {
    super(props);
    this.onLearnMoreClick = this.onLearnMoreClick.bind(this);
    this.toggleMessage = this.toggleMessage.bind(this);
    this.onContextMenu = this.onContextMenu.bind(this);
    this.renderIcon = this.renderIcon.bind(this);
  }

  componentDidMount() {
    if (this.messageNode) {
      if (this.props.scrollToMessage) {
        this.messageNode.scrollIntoView();
      }

      this.emitNewMessage(this.messageNode);
    }
  }

  componentDidCatch(e) {
    this.setState({ error: e });
  }

  // Event used in tests. Some message types don't pass it in because existing tests
  // did not emit for them.
  emitNewMessage(node) {
    const { serviceContainer, messageId, timeStamp } = this.props;
    serviceContainer.emitForTests(
      "new-messages",
      new Set([{ node, messageId, timeStamp }])
    );
  }

  onLearnMoreClick(e) {
    const { exceptionDocURL } = this.props;
    this.props.serviceContainer.openLink(exceptionDocURL, e);
    e.preventDefault();
  }

  toggleMessage(e) {
    // Don't bubble up to the main App component, which  redirects focus to input,
    // making difficult for screen reader users to review output
    e.stopPropagation();
    const { open, dispatch, messageId, onToggle } = this.props;

    // Early exit the function to avoid the message to collapse if the user is
    // selecting a range in the toggle message.
    const window = e.target.ownerDocument.defaultView;
    if (window.getSelection && window.getSelection().type === "Range") {
      return;
    }

    // If defined on props, we let the onToggle() method handle the toggling,
    // otherwise we toggle the message open/closed ourselves.
    if (onToggle) {
      onToggle(messageId, e);
    } else if (open) {
      dispatch(actions.messageClose(messageId));
    } else {
      dispatch(actions.messageOpen(messageId));
    }
  }

  onContextMenu(e) {
    const { serviceContainer, source, request, messageId } = this.props;
    const messageInfo = {
      source,
      request,
      messageId,
    };
    serviceContainer.openContextMenu(e, messageInfo);
    e.stopPropagation();
    e.preventDefault();
  }

  renderIcon() {
    const { level, inWarningGroup, isBlockedNetworkMessage, type } = this.props;

    if (inWarningGroup) {
      return undefined;
    }

    if (isBlockedNetworkMessage) {
      return MessageIcon({
        level: MESSAGE_LEVEL.ERROR,
        type: "blockedReason",
      });
    }

    return MessageIcon({
      level,
      type,
    });
  }

  renderTimestamp() {
    if (!this.props.timestampsVisible) {
      return null;
    }

    return dom.span(
      {
        className: "timestamp devtools-monospace",
      },
      l10n.timestampString(this.props.timeStamp || Date.now())
    );
  }

  renderErrorState() {
    const newBugUrl =
      "https://bugzilla.mozilla.org/enter_bug.cgi?product=DevTools&component=Console";
    const timestampEl = this.renderTimestamp();

    return dom.div(
      {
        className: "message error message-did-catch",
      },
      timestampEl,
      MessageIcon({ level: "error" }),
      dom.span(
        { className: "message-body-wrapper" },
        dom.span(
          {
            className: "message-flex-body",
          },
          // Add whitespaces for formatting when copying to the clipboard.
          timestampEl ? " " : null,
          dom.span(
            { className: "message-body devtools-monospace" },
            l10n.getFormatStr("webconsole.message.componentDidCatch.label", [
              newBugUrl,
            ]),
            dom.button(
              {
                className: "devtools-button",
                onClick: () =>
                  navigator.clipboard.writeText(
                    JSON.stringify(
                      this.props.message,
                      function(key, value) {
                        // The message can hold one or multiple fronts that we need to serialize
                        if (value?.getGrip) {
                          return value.getGrip();
                        }
                        return value;
                      },
                      2
                    )
                  ),
              },
              l10n.getStr(
                "webconsole.message.componentDidCatch.copyButton.label"
              )
            )
          )
        )
      ),
      dom.br()
    );
  }

  // eslint-disable-next-line complexity
  render() {
    if (this.state && this.state.error) {
      return this.renderErrorState();
    }

    const {
      open,
      collapsible,
      collapseTitle,
      source,
      type,
      level,
      indent,
      inWarningGroup,
      topLevelClasses,
      messageBody,
      frame,
      stacktrace,
      serviceContainer,
      exceptionDocURL,
      messageId,
      notes,
    } = this.props;

    topLevelClasses.push("message", source, type, level);
    if (open) {
      topLevelClasses.push("open");
    }

    const timestampEl = this.renderTimestamp();
    const icon = this.renderIcon();

    // Figure out if there is an expandable part to the message.
    let attachment = null;
    if (this.props.attachment) {
      attachment = this.props.attachment;
    } else if (stacktrace && open) {
      attachment = dom.div(
        {
          className: "stacktrace devtools-monospace",
        },
        createElement(SmartTrace, {
          stacktrace,
          onViewSourceInDebugger:
            serviceContainer.onViewSourceInDebugger ||
            serviceContainer.onViewSource,
          onViewSource: serviceContainer.onViewSource,
          onReady: this.props.maybeScrollToBottom,
          sourceMapURLService: serviceContainer.sourceMapURLService,
        })
      );
    }

    // If there is an expandable part, make it collapsible.
    let collapse = null;
    if (collapsible) {
      collapse = createElement(CollapseButton, {
        open,
        title: collapseTitle,
        onClick: this.toggleMessage,
      });
    }

    let notesNodes;
    if (notes) {
      notesNodes = notes.map(note =>
        dom.span(
          { className: "message-flex-body error-note" },
          dom.span(
            { className: "message-body devtools-monospace" },
            "note: " + note.messageBody
          ),
          dom.span(
            { className: "message-location devtools-monospace" },
            note.frame
              ? FrameView({
                  frame: note.frame,
                  onClick: serviceContainer
                    ? serviceContainer.onViewSourceInDebugger ||
                      serviceContainer.onViewSource
                    : undefined,
                  showEmptyPathAsHost: true,
                  sourceMapURLService: serviceContainer
                    ? serviceContainer.sourceMapURLService
                    : undefined,
                })
              : null
          )
        )
      );
    } else {
      notesNodes = [];
    }

    const repeat =
      this.props.repeat && this.props.repeat > 1
        ? createElement(MessageRepeat, { repeat: this.props.repeat })
        : null;

    let onFrameClick;
    if (serviceContainer && frame) {
      if (source === MESSAGE_SOURCE.CSS) {
        onFrameClick =
          serviceContainer.onViewSourceInStyleEditor ||
          serviceContainer.onViewSource;
      } else {
        // Point everything else to debugger, if source not available,
        // it will fall back to view-source.
        onFrameClick =
          serviceContainer.onViewSourceInDebugger ||
          serviceContainer.onViewSource;
      }
    }

    // Configure the location.
    const location = dom.span(
      { className: "message-location devtools-monospace" },
      frame
        ? FrameView({
            frame,
            onClick: onFrameClick,
            showEmptyPathAsHost: true,
            sourceMapURLService: serviceContainer
              ? serviceContainer.sourceMapURLService
              : undefined,
            messageSource: source,
          })
        : null
    );

    let learnMore;
    if (exceptionDocURL) {
      learnMore = dom.a(
        {
          className: "learn-more-link webconsole-learn-more-link",
          href: exceptionDocURL,
          title: exceptionDocURL.split("?")[0],
          onClick: this.onLearnMoreClick,
        },
        `[${l10n.getStr("webConsoleMoreInfoLabel")}]`
      );
    }

    const bodyElements = Array.isArray(messageBody)
      ? messageBody
      : [messageBody];

    return dom.div(
      {
        className: topLevelClasses.join(" "),
        onContextMenu: this.onContextMenu,
        ref: node => {
          this.messageNode = node;
        },
        "data-message-id": messageId,
        "aria-live": type === MESSAGE_TYPE.COMMAND ? "off" : "polite",
      },
      timestampEl,
      MessageIndent({
        indent,
        inWarningGroup,
      }),
      this.props.isBlockedNetworkMessage ? collapse : icon,
      this.props.isBlockedNetworkMessage ? icon : collapse,
      dom.span(
        { className: "message-body-wrapper" },
        dom.span(
          {
            className: "message-flex-body",
            onClick: collapsible ? this.toggleMessage : undefined,
          },
          // Add whitespaces for formatting when copying to the clipboard.
          timestampEl ? " " : null,
          dom.span(
            { className: "message-body devtools-monospace" },
            ...bodyElements,
            learnMore
          ),
          repeat ? " " : null,
          repeat,
          " ",
          location
        ),
        attachment,
        ...notesNodes
      ),
      // If an attachment is displayed, the final newline is handled by the attachment.
      attachment ? null : dom.br()
    );
  }
}

module.exports = Message;
