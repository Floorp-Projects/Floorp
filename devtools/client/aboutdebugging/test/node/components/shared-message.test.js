/* Any copyright is dedicated to the Public Domain.
   http://creativecommons.org/publicdomain/zero/1.0/ */

"use strict";

/**
 * Unit tests for the shared/Message component.
 */

const { shallow } = require("enzyme");
const React = require("react");
const dom = require("devtools/client/shared/vendor/react-dom-factories");

const {
  MESSAGE_LEVEL,
} = require("devtools/client/aboutdebugging/src/constants");

const Message = React.createFactory(
  require("devtools/client/aboutdebugging/src/components/shared/Message")
);

describe("Message component", () => {
  it("renders the expected snapshot for INFO level", () => {
    const message = shallow(
      Message({
        children: dom.div({}, "Message content"),
        className: "some-classname-1",
        level: MESSAGE_LEVEL.INFO,
      })
    );
    expect(message).toMatchSnapshot();
  });

  it("renders the expected snapshot for WARNING level", () => {
    const message = shallow(
      Message({
        children: dom.div({}, "Message content"),
        className: "some-classname-2",
        level: MESSAGE_LEVEL.WARNING,
      })
    );
    expect(message).toMatchSnapshot();
  });

  it("renders the expected snapshot for ERROR level", () => {
    const message = shallow(
      Message({
        children: dom.div({}, "Message content"),
        className: "some-classname-3",
        level: MESSAGE_LEVEL.ERROR,
      })
    );
    expect(message).toMatchSnapshot();
  });
});

describe("Message component renders with closing button", () => {
  it("renders the expected snapshot for Message with closing button", () => {
    const message = shallow(
      Message({
        children: dom.div({}, "Message content"),
        className: "some-classname-1",
        level: MESSAGE_LEVEL.INFO,
        isCloseable: true,
      })
    );
    expect(message).toMatchSnapshot();
  });
});
