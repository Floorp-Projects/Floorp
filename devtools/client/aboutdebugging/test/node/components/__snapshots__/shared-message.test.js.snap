// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Message component renders the expected snapshot for ERROR level 1`] = `
<aside
  className="message message--level-error  qa-message some-classname-3"
>
  <img
    className="message__icon"
    src="chrome://devtools/skin/images/aboutdebugging-error.svg"
  />
  <div
    className="message__body"
  >
    <div>
      Message content
    </div>
  </div>
</aside>
`;

exports[`Message component renders the expected snapshot for INFO level 1`] = `
<aside
  className="message message--level-info  qa-message some-classname-1"
>
  <img
    className="message__icon"
    src="chrome://devtools/skin/images/aboutdebugging-information.svg"
  />
  <div
    className="message__body"
  >
    <div>
      Message content
    </div>
  </div>
</aside>
`;

exports[`Message component renders the expected snapshot for WARNING level 1`] = `
<aside
  className="message message--level-warning  qa-message some-classname-2"
>
  <img
    className="message__icon"
    src="chrome://devtools/skin/images/alert.svg"
  />
  <div
    className="message__body"
  >
    <div>
      Message content
    </div>
  </div>
</aside>
`;

exports[`Message component renders with closing button renders the expected snapshot for Message with closing button 1`] = `
<aside
  className="message message--level-info  qa-message some-classname-1"
>
  <img
    className="message__icon"
    src="chrome://devtools/skin/images/aboutdebugging-information.svg"
  />
  <div
    className="message__body"
  >
    <div>
      Message content
    </div>
  </div>
  <button
    className="ghost-button message__button message__button--info qa-message-button-close-button"
    onClick={[Function]}
  >
    <Localized
      attrs={
        Object {
          "alt": true,
        }
      }
      id="about-debugging-message-close-icon"
    >
      <img
        className="qa-message-button-close-icon"
        src="chrome://devtools/skin/images/close.svg"
      />
    </Localized>
  </button>
</aside>
`;
